File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n=coins.size();
        int t[n+1][amount+1];
        for(int i=0; i<amount+1; i++) t[0][i] = INT_MAX-1;
        for(int i=0; i<n+1; i++) t[i][0] = 0;
        
        for(int i=1; i<amount+1; i++)
            t[1][i] = (i%coins[0])==0? i/coins[0]: INT_MAX-1; 
        
        for(int i=2;i<n+1; i++){
            for(int j=1; j<amount+1; j++){
                if(coins[i-1]<=j) t[i][j] = min(t[i-1][j], 1+ t[i][j-coins[i-1]]);
                else t[i][j] = t[i-1][j];
            }
        }
        
        return t[n][amount]==INT_MAX-1? -1: t[n][amount];
    }
};

File: 0322-coin-change/0322-coin-change.java
class Solution {
    int n;

    int max = Integer.MAX_VALUE-1;
    int [][]memo;
    public int coinChange(int[] coins, int amount) {
        n = coins.length-1;
        memo = new int[n+1][amount+1];

        for(int i=0; i<n+1; i++){
            Arrays.fill(memo[i], -1);
        }
        int ans =  rec(coins, 0, amount);
        return ans == max? -1 : ans;
    }

    private int rec(int[] coins, int idx, int amount){
        if (idx == n){
            if(amount % coins[idx] == 0) return amount/coins[idx];
            return max;
        }

        if(memo[idx][amount] != -1) return memo[idx][amount];

        int notTake = rec(coins, idx+1,amount);

        int take = max;

        if(coins[idx] <= amount){
            take = 1 + rec(coins, idx, amount - coins[idx]);
        }

        return memo[idx][amount] = Math.min(notTake, take);
    }
}